### PR TITLE
fix: boolean query incorrectly dropping documents when AllScorer is present

### PR DESCRIPTION
**Pull request by mdashti**
## What

Fixed a bug where Boolean queries would incorrectly return 0 results (or fewer than expected) when one of the clauses produced an `AllScorer`—typically from range queries matching all documents.

## Why

When `BooleanWeight` processes scorers, it removes `AllScorer` instances as an optimization (since they match everything). However, the code wasn't properly accounting for these removed scorers in certain MUST/SHOULD combinations, causing:

1. **SHOULD + SHOULD with AllScorer**: The "match all" semantics were lost, leaving only the other SHOULD clause
2. **MUST (AllScorer) + SHOULD**: Empty `must_scorers` vector → `intersect_scorers([])` returned `EmptyScorer`
3. **Range queries in Boolean clauses**: When all docs match a range (e.g., `age > 50` when everyone is over 50), the query would silently fail

This commonly surfaced in queries like:
```
(age > 50 OR name = 'Alice') AND status = 'active'
```
when all documents happened to satisfy the range condition.

## How

Introduced two helper functions to centralize the AllScorer handling logic:

**`effective_must_scorer()`** — Returns the effective MUST scorer accounting for removed AllScorers:
- Empty list + had AllScorers → returns `AllScorer` (all docs match)
- Empty list + no AllScorers → returns `None` (no MUST constraint)
- Non-empty list → returns intersection of scorers

**`effective_should_scorer_for_union()`** — Restores "match all" semantics for SHOULD:
- If AllScorers were removed, unions the remaining scorer with `AllScorer`

The `complex_scorer()` match arms now use these helpers consistently:
- **Ignored**: Use effective MUST scorer or empty
- **Optional**: No MUST → promote SHOULD (with union fix); Has MUST → SHOULD only affects scoring
- **Required**: No MUST → SHOULD alone; Has MUST → intersect MUST with SHOULD

## Tests

Added 4 regression tests.

### CODE REVIEW

File: src/query/boolean_query/mod.rs [hunk 1]
```diff
@@ -374,4 +374,240 @@ mod tests {
         }
         Ok(())
     }
+
+    // =========================================================================
+    // AllScorer Preservation Regression Tests
```
Comment (stuhood) [inline src/query/boolean_query/mod.rs:422]: So, it looks like they didn't make it onto `main`, but I had added some property tests to a branch which were intended to cover this code: https://github.com/paradedb/tantivy/blob/2c904e12769eea0d496f8263bfcaad21a9f06858/src/query/boolean_query/mod.rs#L316-L446

I think that @PSeitz decided not to land the branch that those property tests landed on, which is fine. But IMO: that would be a reliable way to add coverage for all of these cases. They don't currently test `AllQuery` explicitly, but I think that that would be a small tweak.

**Review by stuhood — APPROVED** [review]
Thank you!

File: src/query/boolean_query/boolean_weight.rs [hunk 2]
```diff
@@ -97,6 +97,55 @@ fn into_box_scorer<TScoreCombiner: ScoreCombiner>(
...
+/// For union semantics (OR): if any SHOULD clause was an AllScorer, the result
+/// should include all documents. We restore this by unioning with AllScorer.
+fn effective_should_scorer_for_union<TScoreCombiner: ScoreCombiner>(
+    should_scorer: SpecializedScorer,
+    removed_all_scorer_count: usize,
+    max_doc: DocId,
+    num_docs: u32,
+    score_combiner_fn: impl Fn() -> TScoreCombiner,
+) -> SpecializedScorer {
+    if removed_all_scorer_count > 0 {
```
Comment (trinity-1686a) [inline src/query/boolean_query/boolean_weight.rs:138]: this could probably return an AllScorer alone and drop the should_scorer when scoring is disabled

**Comment by trinity-1686a** [comment]
interestingly, it seems it's not possible to trigger this bug with an AllQuery (only 1 of the added tests fails on previous commits, the range one) because it actually generates a `BoostScorer<AllScorer>`, not an AllScorer. It's possible to trigger it with a Range over fastfield, an exist query, and even a TermQuery.

This means we may be missing some optimisations in some case, without knowing it.

I also notice there seems to be places where scoring might be wrong due to AllScorer handling, notably FastFieldRangeWeight doesn't always honor boost (unbounded range on both sides), and BooleanWeight remove most AllScorers without compensating scores. That last one can cause result reordering on some score combiner inside a single segment, and on any combiner when considering multiple segments, where only some segment sees AllScorers, and other see a Range or Exist scorer

File: src/query/boolean_query/mod.rs (hunk 1 above)
Comment (PSeitz) [reply src/query/boolean_query/mod.rs:422]: Actually the opposite, I asked for prop-tests to help land https://github.com/quickwit-oss/tantivy/pull/2538 and they have been merged here:
https://github.com/quickwit-oss/tantivy/pull/2690/files
Extending those tests to cover all/no hits would be nice I think

File: src/query/boolean_query/boolean_weight.rs (hunk 2 above)
Comment (mdashti) [reply src/query/boolean_query/boolean_weight.rs:138]: Good idea. Applied it. Thanks.

File: src/query/boolean_query/mod.rs (hunk 1 above)
Comment (mdashti) [reply src/query/boolean_query/mod.rs:422]: Cherry-picked @stuhood's commit with prop-tests and extended it to cover All and Range.

**Review by mdashti — COMMENTED** [review]
@trinity-1686a @stuhood @PSeitz Thanks for the comments.

> interestingly, it seems it's not possible to trigger this bug with an AllQuery (only 1 of the added tests fails on previous commits, the range one) because it actually generates a BoostScorer<AllScorer>, not an AllScorer. It's possible to trigger it with a Range over fastfield, an exist query, and even a TermQuery.

You're right. Extended the test to better cover Range.

**Comment by mdashti** [comment]
@fulmicoton @PSeitz any more comments?

File: src/query/boolean_query/mod.rs [hunk 3]
```diff
@@ -374,4 +376,413 @@ mod tests {
...
+        /// Matches documents where the field has value "true"
+        Leaf {
+            field_idx: usize,
+        },
+        Union(Vec<BooleanQueryAST>),
+        Intersection(Vec<BooleanQueryAST>),
+    }
+
+    impl BooleanQueryAST {
+        fn matches(&self, doc_id: DocId, num_docs: usize) -> bool {
```
Comment (trinity-1686a) [inline src/query/boolean_query/mod.rs]: `num_docs` feels unused, what is it supposed to convey?

File: src/query/boolean_query/mod.rs [hunk 4]
```diff
@@ -374,4 +376,413 @@ mod tests {
...
+        All,
+        /// Matches all documents via RangeQuery (returns bare AllScorer)
+        /// This is the actual trigger for the AllScorer preservation bug
+        RangeAll,
+        /// Matches documents where the field has value "true"
+        Leaf {
+            field_idx: usize,
+        },
+        Union(Vec<BooleanQueryAST>),
+        Intersection(Vec<BooleanQueryAST>),
```
Comment (trinity-1686a) [inline src/query/boolean_query/mod.rs:703]: ideally i think we'd want a `Vec<(Occur, BooleanQueryAST)>` to test the edge cases that can happen when mixing must and should. But some proptest is already good to have, and we can improve that later

File: src/query/boolean_query/mod.rs [hunk 5]
```diff
@@ -374,4 +376,413 @@ mod tests {
...
+        // Leaf strategies: term queries, AllQuery, and RangeQuery matching all docs
+        let leaf = prop_oneof![
+            (0..num_fields).prop_map(|field_idx| BooleanQueryAST::Leaf { field_idx }),
+            Just(BooleanQueryAST::All),
+            Just(BooleanQueryAST::RangeAll),
+        ];
+        leaf.prop_recursive(
+            8,   // 8 levels of recursion
+            256, // 256 nodes max
+            10,  // 10 items per collection
```
Comment (trinity-1686a) [inline src/query/boolean_query/mod.rs:806]: i think these values are too high, and will lead to a very large search space with many redundant elements.
`3, 128, 8` feels like probably enough

**Review by trinity-1686a — APPROVED** [review]
i tried to poke hole in your changes, and couldn't find anything unexpected, except for [an issue](https://github.com/quickwit-oss/tantivy/issues/2773) that was already present before. Can you just clarify what `num_docs` is supposed to do in `BooleanQueryAST::matches` (and remove it if it's really not used) before i merge?

File: src/query/boolean_query/mod.rs [hunk 6]
```diff
@@ -374,4 +376,413 @@ mod tests {
...
+                        .iter()
+                        .map(|child| (Occur::Must, child.to_query(fields, range_field)))
+                        .collect();
+                    Box::new(BooleanQuery::new(sub_queries))
+                }
+            }
+        }
+    }
+
+    fn create_index_with_boolean_permutations(num_fields: usize) -> (Index, Vec<Field>, Field) {
```
Comment (stuhood) [inline src/query/boolean_query/mod.rs]: It looks like you removed the `num_docs` parameter while cherry-picking this, but that turned out to be important for reproducing problems with certain optimizations around batching/blocking: https://github.com/quickwit-oss/tantivy/pull/2697

**Comment by fulmicoton** [comment]
@stuhood Can you rebase/fix the conflict?

File: src/query/boolean_query/mod.rs (hunk 5 above)
Comment (stuhood) [reply src/query/boolean_query/mod.rs:806]: Possibly... but the tests run at about 8 microseconds per case (so sub- 250 ms for the default 256 cases), so it doesn't seem like we need to prune here unless these get a lot slower?

**Comment by stuhood** [comment]
> @stuhood Can you rebase/fix the conflict?

Done. Thanks!

File: src/query/boolean_query/mod.rs (hunk 5 above)
Comment (trinity-1686a) [reply src/query/boolean_query/mod.rs:806]: the goal wouldn't be to improve perf, but to make the search space smaller, while hopefully keeping it big enough that it still contains instances of any interesting edge-cases. This makes it more likely to find something interesting.

As a made up example, if you test a function that evaluates to `f(x) = 1/(1+x)` randomly 256 times over the range `[-10^8; 10^8]`, you're unlikely to discover it can fail. Doing the same over `[-10^3; 10^3]`, you have a reasonable chance of discovering the edge case

Comment (stuhood) [reply src/query/boolean_query/mod.rs:806]: > As a made up example, if you test a function that evaluates to `f(x) = 1/(1+x)` randomly 256 times over the range `[-10^8; 10^8]`, you're unlikely to discover it can fail. Doing the same over `[-10^3; 10^3]`, you have a reasonable chance of discovering the edge case

I'm not sure that this case is comparable. Rather than choosing numbers in a distribution, we're choosing a tree structure. For example, if it takes a parent of type `X`, and a child of type `Y` to reproduce a particular bug, then the depth of that parent and child in the tree aren't particularly relevant, because the framework will shrink the repro as much as possible once it finds it.

We're choosing 256 items from a choice of N enums... so rather than choosing a single value in a huge space like `[-10^8; 10^8]`, this is more like choosing 256 numbers in the space of `[-10^3; 10^3]` (and then doing that 256 times, by default), some permutations of which might fail.

**Comment by fulmicoton** [comment]
Thank you!
